### DISCUSSION TITLE
What special tokens to use for transcription, and why is <|startoflm|> so highly probable?

### DISCUSSION THREAD

**Discussion by belson17** [openai/whisper#2578] [Q&A]
Hi all, I'm trying to determine the probability of a sequence of "target tokens" for a given audio recording of speech. To do so, I'm passing in this sequence of tokens and the encoded Mel spectrogram to the decoder. My problem is I'm unsure what special tokens to prepend to the sequence to indicate to the model to transcribe in English. In particular, for the "turbo" and "large-v3" models, I see the token "<|startoflm|>" has a very high probability output from the decoder, so it seems the model expects this token, but I don't know how to use it. I also don't see this token predicted with high probability when using other models, e.g., "base". 

What prefix special tokens should I use for the turbo and large models? Below are snippets of code to show what I'm doing. If there is documentation, please point me there. Thanks!

```
model = whisper.load_model("turbo", device="cuda")

prefix_token_ids = [50258, 50259, 50359, 50363]
prefix_tokens = ['<|startoftranscript|>', '<|en|>', '<|transcribe|>', '<|notimestamps|>']
prefix_tensor = torch.tensor(prefix_token_ids).unsqueeze(0)

full_sequence_token_ids = torch.cat([prefix_tensor, target_token_ids], dim=1)
decoder_input_token_ids = full_sequence_token_ids[:, :-1]
target_token_ids_to_eval = full_sequence_token_ids[:, 1:]

logits = model.decoder(decoder_input_token_ids, self.model.encoder(mel))
log_probs = F.log_softmax(logits, dim=-1)

for i in range(len(logits[0]):
    pred_top = logits[0, i].topk(5).indices
    pred_top_probs = np.exp(log_probs[0, i, pred_top].detach().cpu().numpy())
    print(f"Time step {i}: top tokens = {[tokenizer.decode([t]) for t in pred_top]}, probs={pred_top_probs}")
    print(f"Target token: {self.tokenizer.decode([target_token_ids_to_eval[0, i]])}")
```

Outputs:
```
Time step 0: top tokens = ['<|en|>', '<|notimestamps|>', '<|la|>', '<|de|>', '<|fr|>'], probs=[9.9240476e-01 4.9215057e-03 4.9104204e-04 2.9613255e-04 2.5190943e-04]
Target token: <|en|>
Time step 1: top tokens = ['<|startoflm|>', '<|transcribe|>', '', '<|nospeech|>', '<|endoftext|>'], probs=[9.9999833e-01 1.6491989e-06 2.2530656e-08 1.3922099e-08 4.4615808e-10]
Target token: <|transcribe|>
Time step 2: top tokens = ['', '', '', '', ''], probs=[5.7725841e-01 4.1457579e-01 5.2851834e-04 5.1052548e-04 4.3363243e-04]
Target token: <|notimestamps|>
Time step 3: top tokens = ['<|endoftext|>', ',', ' and', ' -', ' ...'], probs=[1.0000000e+00 3.1554283e-11 2.2344248e-11 1.2954151e-11 1.1298531e-11]
Target token:  They
Time step 4: top tokens = [' regain', ' reg', ' began', ' re', ' begin'], probs=[0.6244507  0.309043   0.00685263 0.00403515 0.00258522]
Target token:  regain
```

**Comment by Advait251206** [comment]
What you're seeing is expected, and it stems from an architectural difference between **Large-v3/Turbo** and earlier Whisper models.

## Short answer

For transcription with **Large-v3** and **Turbo**, you should generally use the same transcription prefix that Whisper itself uses:

```text
<|startoftranscript|>
<|en|>
<|transcribe|>
<|notimestamps|>
```

However, if you're manually computing sequence probabilities by calling:

```python
model.decoder(...)
```

you may be missing an additional token that the decoder expects internally, which is why you're seeing:

```text
P(<|startoflm|>) ≈ 0.999998
```

at position 1.

## Why is `<|startoflm|>` so probable?

The key observation is:

```text
Input:
<|startoftranscript|>

Prediction:
<|en|>
```

which looks correct.

Then:

```text
Input:
<|startoftranscript|> <|en|>

Prediction:
<|startoflm|>
```

instead of:

```text
<|transcribe|>
```

This suggests that the decoder's training distribution for Large-v3/Turbo includes an intermediate language-modeling token that your manually constructed sequence is not providing.

In other words:

```text
Your prefix:
<|startoftranscript|>
<|en|>
<|transcribe|>
<|notimestamps|>
```

may not exactly match the sequence used during training or generation.

## Important distinction: decoder vs generate()

When Whisper runs normally:

```python
model.decode(...)
```

or

```python
model.transcribe(...)
```

it constructs the prompt using tokenizer logic and decoding rules.

When you directly call:

```python
model.decoder(token_ids, audio_features)
```

you bypass that machinery and become responsible for reproducing the exact token sequence expected by the model.

This becomes particularly important for:

- Large-v3
- Turbo

because they introduced tokenizer and decoder changes compared to older checkpoints.

## Check what Whisper actually uses

I would inspect:

```python
tokenizer = whisper.tokenizer.get_tokenizer(
    model.is_multilingual,
    language="en",
    task="transcribe"
)
```

and then examine:

```python
tokenizer.sot_sequence
```

and related attributes.

For example:

```python
print(tokenizer.sot_sequence)
```

You may find that the actual decoding prompt differs from the four-token prefix you're manually constructing.

## Another clue: your probabilities collapse

You observe:

```text
Time step 3:
<|endoftext|> = 1.0
```

before the model has generated meaningful text.

That's a strong signal that the decoder believes the sequence is malformed.

If the prefix were correct, you'd expect:

```text
They
The
I
We
...
```

to appear among the most likely next tokens.

Instead, the model appears to think:

```text
sequence finished
```

or

```text
invalid context
```

which often happens when a required control token is missing.

## How I'd debug this

Compare the logits from:

```python
model.decode(...)
```

with those from:

```python
model.decoder(...)
```

for the same audio.

Specifically:

1. Run a normal decode.
2. Capture the exact token sequence being fed to the decoder.
3. Use that sequence in your probability computation.

That removes any ambiguity about special tokens.

## About `<|startoflm|>`

From experimentation and community reports, Large-v3/Turbo appear to use some internal control-token behavior that differs from earlier models.

The fact that:

```text
P(<|startoflm|>) ≈ 0.999998
```

implies the model strongly expects that token at that position under the context you've provided.

Rather than manually guessing where it belongs, I'd recommend extracting the actual prompt sequence generated by Whisper's tokenizer and decoding pipeline for those checkpoints.

## Recommendation

Instead of hardcoding:

```python
[50258, 50259, 50359, 50363]
```

use Whisper's tokenizer utilities to obtain the exact start-of-transcript sequence for the model you're evaluating.

If your goal is sequence likelihood evaluation, the safest approach is:

```python
tokenizer = get_tokenizer(...)
prompt = tokenizer.sot_sequence
```

and then append your target text tokens exactly as Whisper would.

The unusually high probability of `<|startoflm|>` is a strong indication that the manually specified prefix does not fully match the decoding prompt expected by Large-v3/Turbo, even though it works reasonably for earlier models such as `base`.

**Reply by belson17** [reply]
This reply reads like AI slop to me.

**Reply by Advait251206** [reply]
A little bit

**Comment by Samielakkad** [comment]
`<|startoflm|>` is not an extra token that Large-v3/Turbo require here. The result comes from mixing a 100-language model with hard-coded IDs (and apparently a tokenizer) for the older 99-language vocabulary.

Whisper inserts all language tokens *before* `<|translate|>`, `<|transcribe|>`, and the remaining control tokens in [`get_encoding()`](https://github.com/openai/whisper/blob/main/whisper/tokenizer.py#L337-L351). The model exposes the matching language count as [`model.num_languages`](https://github.com/openai/whisper/blob/main/whisper/model.py#L307-L308), and the normal decoder passes that value to `get_tokenizer()`.

That means the control-token IDs after the language block shift by one:

```text
99-language models:  transcribe=50359, no_timestamps=50363
100-language models: transcribe=50360, no_timestamps=50364
```

So `50360` is the model's `<|transcribe|>` for Large-v3/Turbo, but a 99-language tokenizer decodes the same integer as `<|startoflm|>`. That exactly explains the apparently near-certain `<|startoflm|>` prediction.

Do not hard-code the prefix. Build it with the model's vocabulary metadata and use the same tokenizer for the target text and for decoding logits:

```python
from whisper.tokenizer import get_tokenizer

tokenizer = get_tokenizer(
    model.is_multilingual,
    num_languages=model.num_languages,
    language="en",
    task="transcribe",
)

prefix_token_ids = list(tokenizer.sot_sequence_including_notimestamps)
prefix_tensor = torch.tensor(prefix_token_ids, device=model.device).unsqueeze(0)
```

For Large-v3/Turbo that produces the model-aligned sequence instead of `[50258, 50259, 50359, 50363]`. Your next-token shifting (`inputs = full[:-1]`, `targets = full[1:]`) is otherwise the right shape for scoring a fixed sequence.
